src/core/qgssnappingutils.cpp

```diff
@@ -92,8 +94,44 @@ bool QgsSnappingUtils::isIndexPrepared( QgsVectorLayer *vl, const QgsRectangle &
   return ( mStrategy == IndexHybrid || mStrategy == IndexExtent ) && loc->hasIndex() && ( !loc->extent() || loc->extent()->contains( aoi ) ); // the index - even if it exists - is not suitable
 }
 
+static bool _isMatchAVisibleLayer( const QgsPointLocator::Match &candidateMatch, QgsMapSettings mapSettings, bool snapOnInvisibleFeature )
+{
+
+  bool visible = true;
+  if ( !snapOnInvisibleFeature )
+  {
+    // segmentIntersection haven't a layer
+    if ( candidateMatch.layer() )
+    {
+      bool filter = false;
+      QgsRenderContext context( QgsRenderContext::fromMapSettings( mapSettings ) );
+      context.expressionContext() << QgsExpressionContextUtils::layerScope( candidateMatch.layer() );
+      std::unique_ptr< QgsFeatureRenderer > renderer( candidateMatch.layer()->renderer() ? candidateMatch.layer()->renderer()->clone() : nullptr );
+      if ( renderer )
+      {
+        // setup scale for scale dependent visibility (rule based)
+        renderer->startRender( context, candidateMatch.layer()->fields() );
+        filter = renderer->capabilities() & QgsFeatureRenderer::Filter;
+      }
+
+      QgsFeature feat = candidateMatch.layer()->getFeature( candidateMatch.featureId() );
```

Comment:
This is going to be a performance killer too - can you find a way to avoid fetching individual features and instead do a single feature fetch for the whole canvas extent at once?